enhancements/ocp-coreos-layering.md

```diff
@@ -0,0 +1,496 @@
+
+The OCP admins at example.corp get a copy of this hotfix RPM into their internal data store, and configure a derived build where the kernel RPMs are applied. The MCO builds a derived image and rolls it out.
+
+(Note: this flow would likely be explained as a customer portal document, etc.)
+
+Later, a fixed kernel with a newer version is released in the main OCP channels.  The override continues to apply, and will currently require manual attention at some point later to remove the override when we can guarantee the non-hotfix kernel has the desired fix.
+
+A future enhancement will help automate the above.  For example, we may support associating a Bugzilla number with the override.  Then, the system can more intelligently look at an upgrade and see whether a new kernel package that is chronologically/numerically higher actually fixes the BZ - and only drop the override once it does.
+
+#### Externally built image
+
+As we move towards having users manage many clusters (10, 100 or more), it will make sense to support building a node image centrally.  This will allow submitting the image to a security scanner or review by a security team before deploying to clusters.
+
+Acme Corp has 300 clusters distributed across their manufacturing centers.  They want to centralize their build system in their main data center, and just distribute those images to single node edge machines.  They provide a `custom-coreos-imagestream` object at installation time, and their node CoreOS image is deployed during the installation of each cluster without a build operation.
+
+In the future, we may additionally support a signature mechanism (IMA/fs-verity) that allows these images can be signed centrally.  Each node verifies this image rather than generating a custom state.
+
+(Note some unanswered questions below)
+
+### API Extensions
+
+We will continue to support MachineConfig.  The exact details of the mechanics of custom builds are still TBD.
+
+### Implementation details
+
+<details>
+
+**(NOTE!  These details are provisional and we intend a forthcoming enhancement to detail more)**
+
+#### Upgrade flow
+
+See [OSUpgrades](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md) for the current flow before this proposal.
+
+1. The "base image" will be part of the release image, the same as today.
+1. The CVO will replace a ConfigMap in the MCO namespace with the OS payload reference, as it does today.
+1. The MCO will update an `imagestream` object (e.g. `openshift-machine-config-operator/rhel-coreos`) when this ConfigMap changes.
+1. The MCO builds a container image for each `MachineConfigPool`, using the new base image
+1. Each machineconfig pool will also support a `custom-coreos` `BuildConfig` object and imagestream.  This build *must* use the `mco-coreos` imagestream as a base.   The result of this will be rolled out by the MCO to nodes.
```

Comment:
so if i'm understanding it correctly, this is the spot at which a user/admin would define the image customizations they want to apply, by defining a buildconfig that points to a dockerfile that contains their additions, more or less?

picking up @fabiand's question from the mailing list:

> How is it imagined to include different packages from different vendors in the same base image?
Let's say we've got a HW vendor providing (proprietary) drivers for their GPU, hardware monitoring tools by the system vendor, and some IDS. Some of this will be provided by an Operator, other elements by a human. How are they expected to create a single new base image?

and @cgwalters answer:
> Hmm; in theory perhaps we could create an interface (like a CRD) that
allowed registering "build fragments" (much like MachineConfig) that
could all be gathered and merged into a single build.

is there another option here that we allow multiple user-provided buildconfigs, with some sort of ordering, where only the final build produces the actual coreos image that will be consumed by the cluster?  I don't know if it's strictly necessary, vs just aggregating all the customization logic into a single buildconfig/dockerfile, but i wanted to continue the thread.

Reply:
I feel confident that: no matter what we pick, it will have feature gaps :smile:

This enhancement specifically didn't yet try to call out the details of how we propose the user interface to work, because I think we simply need time to iterate on it, provide a functioning preview and and learn what works before we can draw a roadmap into that area.

Reply:
totally fair, just trying to get some thoughts in place for when you get to that point